zookeeper-server/src/main/java/org/apache/zookeeper/ClientCnxn.java

```diff
@@ -855,10 +855,22 @@ class SendThread extends ZooKeeperThread {
         private boolean isFirstConnect = true;
         private volatile ZooKeeperSaslClient zooKeeperSaslClient;
 
-        private String stripChroot(String serverPath) {
+        String stripChroot(String serverPath) {
             if (serverPath.startsWith(chrootPath)) {
                 if (serverPath.length() == chrootPath.length()) {
                     return "/";
+                } else if (serverPath.charAt(chrootPath.length()) != '/') {
+                    // Corner-case:
+                    // * Event for config node "/zookeeper/config".
+                    // * Chroot "/zoo".
+                    //
+                    // We can't simply deliver "/zookeeper/config" in all cases, as
+                    // getData("/config") in chroot "/zookeeper" expect "/config".
+                    // But at least, we should not deliver abnormal path here.
+                    //
+                    // It might be better to move chroot out of here as ZOOKEEPER-838
+                    // suggested, but it is another story.
+                    return serverPath;
                 }
                 return serverPath.substring(chrootPath.length());
             } else if (serverPath.startsWith(ZooDefs.ZOOKEEPER_NODE_SUBTREE)) {
```

Comment:
I wonder if we can simply switch the if conditions order. That is:

```java
if (serverPath.startsWith(ZooDefs.ZOOKEEPER_NODE_SUBTREE)) {
  return serverPath;
} else if (serverPath.startsWith(chrootPath)) {
  // ...
}
```

Reply:
But still, the original change @kezhuw propose if reasonable that we can prevent malformed path.

Reply:
@tisonkun Thank you for take a look at this!

`getData("/config", watcher, stat)` in chroot "/zookeeper" should receive `/config` for its `watcher`. Simple reversing the order does not satisfy above.

This pr concentrate more on `getConfig` which should receive `/zookeeper/config` for its watcher in any cases. 

The above code is a leftover from [ZOOKEEPER-4565](https://issues.apache.org/jira/browse/ZOOKEEPER-4565)(#1899). I could split it to a separate issue if you prefer. It could possibly make this pr more concentrated. 

> In chroot "/zoo", the watcher receives abnormal path "keeper/config". [ZOOKEEPER-4565](https://issues.apache.org/jira/browse/ZOOKEEPER-4565) should fix this, but failed to do so.